### PR TITLE
Handle kubectl ... --watch command

### PR DESCRIPTION
**Pull request by sidpremkumar**
## Description of the change

Adds support to proxy `kubectl ... --watch` commands which is needed in order to get lens working. 

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Refactor (code change with no functionality change)
- [ ] Enhancement (minor change below the level of a feature)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Remove feature (removes a feature we don't support anymore)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related JIRA tickets

Relates to JIRA: CWC-1113

## Relevant release note information

Release Notes:

## Potential Security Impacts

Does this PR have any security impact?
Does it fix a security issue?
Does it add attack surface?
Why do we think this change is safe?

### Development

- [ ] Does the code changed/added as part of this pull request have test coverage?
- [ ] Do all tests related to the changed code pass in development?
- [ ] Have you ensured that at least one other person has tested your code?
- [ ] Have you tested the code?
- [ ] Have you attached any pictures (if relevant)?

### CODE REVIEW

File: kube-go-assets/bctl/agent/plugin/kube/actions/watch/watch.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+		l.closed = true
+		return string(WatchStop), []byte{}, nil
+	default:
+		rerr := fmt.Errorf("unhandled watch action: %v", action)
+		l.logger.Error(rerr)
+		return "", []byte{}, rerr
+	}
+}
+
+func (l *WatchAction) validateRequestId(requestId string) error {
```
Comment (sidpremkumar) [inline kube-go-assets/bctl/agent/plugin/kube/actions/watch/watch.go:95]: lets try to move as much of this as possible to `utils.go`

File: kube-go-assets/bctl/daemon/plugin/kube/actions/watch/watch.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+			var kubewatchHeadersPayload kubewatch.KubeWatchHeadersPayload
+			if err := json.Unmarshal(contentBytes, &kubewatchHeadersPayload); err != nil {
+				// If we see an error this must be an early message
+				earlyMessages[earlyMessageNumber] = contentBytes
+				earlyMessageNumber += 1
+			} else {
+				// This is our header message, loop and apply
+				for name, values := range kubewatchHeadersPayload.Headers {
+					for _, value := range values {
+						w.logger.Info(fmt.Sprintf("HERE: %s", value))
```
Comment (sidpremkumar) [inline kube-go-assets/bctl/daemon/plugin/kube/actions/watch/watch.go:100]: remove this

File: kube-go-assets/bctl/daemon/plugin/kube/kube.go [hunk 3]
```diff
@@ -270,6 +283,42 @@ func (k *KubeDaemonPlugin) rootCallback(w http.ResponseWriter, r *http.Request)
...
+	if follow[0] == "true" || follow[0] == "1" {
+		return true
+	}
+
+	// Else return false
+	return false
+}
+
+func isWatchRequest(request *http.Request) bool {
+	// Determin if we are trying to watch the resource
```
Comment (lgmugnier) [inline kube-go-assets/bctl/daemon/plugin/kube/kube.go:305]: Determin -> Determine

File: kube-go-assets/bctl/daemon/plugin/kube/actions/watch/watch.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	}
+
+	payloadBytes, _ := json.Marshal(payload)
+	w.RequestChannel <- plgn.ActionWrapper{
+		Action:        startLogs,
+		ActionPayload: payloadBytes,
+	}
+
+	// Wait for our initial message to determine what headers to use
+	earlyMessages := make([][]byte, kubewatch.WatchBufferSize)
```
Comment (lgmugnier) [inline kube-go-assets/bctl/daemon/plugin/kube/actions/watch/watch.go:80]: What are these early messages? Why don't they show up in logs too?

File: kube-go-assets/bctl/daemon/plugin/kube/actions/watch/watch.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	// If there are any early messages, stream them first
+	for _, earlyMessage := range earlyMessages {
+		err := w.writeToHttpRequest(earlyMessage, writer)
+		if err != nil {
+			return nil
+		}
+	}
+
+	// Now subscribe to the response
+	// Keep this as a non-go function so we hold onto the http request
```
Comment (lgmugnier) [inline kube-go-assets/bctl/daemon/plugin/kube/actions/watch/watch.go:118]: I think "function" is supposed to be "routine" here?

**Comment by lgmugnier** [comment]
Did you make watch different from logs because you want to handle them differently at the bastion or because they're different protocols?

**Review by lgmugnier — APPROVED** [review]

**Comment by sidpremkumar** [comment]
Merging this now so Lucie can get started on her work. Planning on making another PR here to generalize the code between logs and watch.

**Comment by sidpremkumar** [comment]
Merging this now so Lucie can get started on her work. Planning on making another PR here to generalize the code between logs and watch.

**Pull request merged**